core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java

```diff
@@ -710,20 +711,34 @@ void checkTaskTimeout(TaskDef taskDef, TaskModel task) {
         }
 
         long timeout = 1000L * taskDef.getTimeoutSeconds();
+        long totalTaskTimeout = 1000L * taskDef.getTotalTimeoutSeconds();
         long now = System.currentTimeMillis();
         long elapsedTime =
                 now - (task.getStartTime() + ((long) task.getStartDelayInSeconds() * 1000L));
 
-        if (elapsedTime < timeout) {
-            return;
+        long elapsedTimeFromFirstTaskExecution =
+                now - (task.getFirstStartTime() + ((long) task.getStartDelayInSeconds() * 1000L));
+
+        if (totalTaskTimeout > 0) {
+            // If totalTaskTimeout is greater than 0, check both conditions
+            if (elapsedTimeFromFirstTaskExecution < totalTaskTimeout && elapsedTime < timeout) {
+                return;
+            }
+        } else {
+            // If totalTaskTimeout is not greater than 0, only check the timeout condition
+            if (elapsedTime < timeout) {
+                return;
+            }
         }
 
         String reason =
                 String.format(
                         "Task timed out after %d seconds. Timeout configured as %d seconds. "
+                                + "Total Task Timeout configured as %d seconds. "
```

Comment:
This `reason` may be confusing. 

It's not immediately clear which time out kicked in since it's mentioning both (timeout and total timeout).